### PR TITLE
#11869 Configure PyPi trusted publishing.

### LINKED ISSUE #11869: New way to publish to PyPI
**Is your feature request related to a problem? Please describe.**
PyPI has created a new and more secure way to publish packages. It's called [Trusted Publishing](https://blog.pypi.org/posts/2023-04-20-introducing-trusted-publishers/) and it's more secure because it uses short-lived tokens instead of long-lived tokens. Currently, twisted is published using a long-lived token, stored in a GitHub secret, in [`test.yml`](https://github.com/twisted/twisted/blob/trunk/.github/workflows/test.yaml#L375-L379) workflow.

**Describe the solution you'd like**
I'd like us to delete the `PYPI_UPLOAD_TOKEN` secret and make a small change in `test.yml` workflow to do Trusted Publishing.

**Describe alternatives you've considered**
None.

**Additional context**
Hi! I'm Gabriela and I work on behalf of Google and the OpenSSF suggesting supply-chain security changes :)

### PR DESCRIPTION
**Pull request by graingert**
## Scope and purpose

Fixes #11869

This splits the package build and publish steps, moves the publish steps into a GHA environment that only allows running on tags

We should add tag protection rules

## Contributor Checklist:

This process applies to *all* pull requests - no matter how small.
Have a look at [our developer documentation](https://docs.twisted.org/en/latest/core/development/dev-process.html) before submitting your Pull Request.

Below is a non-exhaustive list (as a reminder):

* The title of the PR should describe the changes and starts with the associated issue number, like “#1234 Brief description”.
* A release notes news fragment file was create in src/twisted/newsfragments/ (see: [Release notes fragments docs.](https://docs.twisted.org/en/latest/core/development/dev-process.html#release-notes-management))
* The automated tests were updated.
* Once all checks are green, request a review by leaving a comment that contains exactly the string `please review`.
  Our bot will trigger the review process, by applying the pending review label
  and requesting a review from the Twisted dev team.

### CODE REVIEW

File: .github/workflows/test.yaml
Comment (graingert) [inline .github/workflows/test.yaml]:
```suggestion
  # Used exclusively for various release publishing.
  # This is only executed on tags, but just does the publish to exercise the release as much
  # as possible and reduce the possibility of finding bugs in the release
  # process late in the release cycle,
```

File: .github/workflows/test.yaml
Comment (graingert) [inline .github/workflows/test.yaml:493]:
```suggestion
  # Used for various release automation, but not publishing
  # This is also executed for each PR to exercise the release as much
  # as possible and reduce the possibility of finding bugs in the release
  # process late in the release cycle,
  # The files are published only when a tag is created using the next job, release-publish
  release-build:
      name: Build and check release, then upload artifact
```

File: .github/workflows/test.yaml
Comment (graingert) [inline .github/workflows/test.yaml]:
```suggestion
```

File: .github/workflows/test.yaml
Comment (graingert) [inline .github/workflows/test.yaml]:
```suggestion
```

File: .github/workflows/test.yaml
Comment (graingert) [inline .github/workflows/test.yaml]:
```suggestion
        if: ${{ contains(['success', 'skipped'], needs.release-publish.result) }}
```

File: .github/workflows/test.yaml
Comment (graingert) [inline .github/workflows/test.yaml]:
```suggestion
        if: ${{ contains(fromJSON('["success", "skipped"]'), needs.release-publish.result) }}
```

File: .github/workflows/test.yaml
Comment (graingert) [inline .github/workflows/test.yaml]:
```suggestion
        if: ${{ !contains(fromJSON('["success", "skipped"]'), needs.release-publish.result) }}
```

File: .github/workflows/test.yaml
Comment (graingert) [inline .github/workflows/test.yaml]:
```suggestion
  release-publish-success-or-skipped:
```

File: .github/workflows/test.yaml
Comment (graingert) [inline .github/workflows/test.yaml]:
```suggestion
      - release-publish-success-or-skipped
```

File: .github/workflows/test.yaml
Comment (adiroiban) [inline .github/workflows/test.yaml:547]: can we skip this step, instead of the whole job?

In this way, we should no longer need ` release-publish-success-or-skipped` ?

File: .github/workflows/test.yaml
Comment (adiroiban) [inline .github/workflows/test.yaml]: Maybe explain why we have the separate job for publishing.

```suggestion
  # Used for various release automation, but not for publishing, as publishing needs extra permissions for the API token.
  # So we isolate the actual publish step in a separate job.
```

File: .github/workflows/test.yaml
Comment (graingert) [reply .github/workflows/test.yaml:547]: No, because the environment is not available to branches, we have to skip the whole job otherwise the unavailable environment will fail the job and not get to the point of running any steps

Comment (adiroiban) [reply .github/workflows/test.yaml:547]: I missed the GitHub environment part.

Do we need an environment ?

I feel that we are adding a lot of complexity and will make it harder for future maintainer to know what is going on there.

I think that this is the first place where we are introducing GItHub Action Environments in our CI.

If this is the first place, we need to add some documentation about how we plan to use them and how to configure them

I see a setting at https://github.com/twisted/twisted/settings/environments

Comment (graingert) [reply .github/workflows/test.yaml:547]: An environment is strongly recommended, and helps prevent accidents

Comment (adiroiban) [reply .github/workflows/test.yaml:547]: I know that in big companies, where you have a lot of developers that have no idea what they are doing, environments are of great help.

but I think that for the Twisted team, we can live with a bit of risk :)

I am not -1 in terms of environments, but we need to document them and update the release process, so that the person who will try to do a release will know which "wheel" are spinning during the release process

Comment (graingert) [reply .github/workflows/test.yaml:547]: Happy to add a document for this

Comment (graingert) [reply .github/workflows/test.yaml:547]: I've added some notes to the release process docs

**Comment by adiroiban** [comment]
Thanks. I think that we can merge this PR and then test it on the release branch

There is a small error with the docs

> /home/runner/work/twisted/twisted/docs/development/release-process.rst:322: WARNING: Inline emphasis start-string without end-string.


The release docs updates can be reviewed here 

https://twisted--12379.org.readthedocs.build/en/12379/development/release-process.html#notes


--------

The current environment configuration is

image


I have added a requirements for PR approval from a member of the team

File: docs/development/release-process.rst
Comment (adiroiban) [inline docs/development/release-process.rst]:
```suggestion
The release process uses the `release <https://github.com/twisted/twisted/settings/environments/4731362866/edit>`_ GitHub Actions environment.
```

File: docs/development/release-process.rst
Comment (adiroiban) [inline docs/development/release-process.rst]: Maybe have a more explicit title
 
```suggestion
Secirity notes
--------------
```

File: docs/development/release-process.rst
Comment (adiroiban) [inline docs/development/release-process.rst:328]: I don't know what this means.

I think it's important to note that that the `test.yaml` GitHub Worflow is approved to trigger a release for PyPI.

If in the future we do a release from a workflow with a different name, we need to update the PyPI security configuration.

File: docs/development/release-process.rst
Comment (adiroiban) [inline docs/development/release-process.rst]: I think that we should also mention  the PyPI side of this configuration.

```suggestion
#. Github Actions will upload the dist to PyPI when a new tag is pushed to the repo, using the GitHub 'release' environment.
#. In PyPI the GitHub Actions `test.yaml` workflow is configure to allow publishing new PyPI releases.
```

File: docs/development/release-process.rst
Comment (graingert-coef) [inline docs/development/release-process.rst:322]:
```suggestion
The release process uses a GitHub Actions environment, configured here:
```

File: .github/workflows/test.yaml
Comment (adiroiban) [inline .github/workflows/test.yaml]: Do we need this extra job ?

To keep it simple, I was thinking that we can have `release-publish` as an "optional" steps.

We can keep only `release-build` here

It will be skipped on all bracnhes...and we can check that it is executed for a tag.

File: docs/development/release-process.rst
Comment (graingert) [inline docs/development/release-process.rst]:
```suggestion
The release process uses a GitHub Actions environment, configured here:
```

File: .github/workflows/test.yaml
Comment (adiroiban) [inline .github/workflows/test.yaml]: I hope that we can have this implemented without this extra job.

If we need to keep this job, I think that this needs a comment to explai why we need it.

**Review by adiroiban — APPROVED** [review]
Thanks. It looks good. Only minor comments.

Looking forward to merge it and then test it on the release branch

File: docs/development/release-process.rst
Comment (graingert) [reply docs/development/release-process.rst:328]: But that doesn't work with the all-successful pattern, we should keep this test.yaml forever. If we need to split up the file we can import workflows into it

Comment (adiroiban) [reply docs/development/release-process.rst:328]: I wanted to add this mention as I remember that in another project there was a rename from `test.yml` to `test.yaml` and it was confusing why PyPI publishing stopped working

Comment (graingert) [reply docs/development/release-process.rst:328]: Ah ok, well do you want to add a screenshot of the config here?

File: .github/workflows/test.yaml
Comment (graingert) [reply .github/workflows/test.yaml]: We need this so if a publish fails it's detected by the all-successful step

File: docs/development/release-process.rst
Comment (adiroiban) [reply docs/development/release-process.rst:328]: I don't think that we should add the screenshot to the docs.

It's enough to have the screenshot on the PR, as a snapshot of the configuration at the time this was merged

File: .github/workflows/test.yaml
Comment (adiroiban) [reply .github/workflows/test.yaml]: But we don't run this job for a PR ? 
The `all-success` job is designed to prevent merging a PR with failed jobs.


------

Also, the job is triggered only for a tag... so even if it fails. there is not much we can do, other than create a new tag ... or delete and recreate the tag

File: docs/development/release-process.rst
Comment (adiroiban) [inline docs/development/release-process.rst]:
```suggestion
The release process uses a GitHub Actions environment, configured `here
<https://github.com/twisted/twisted/settings/environments/4731362866/edit>`_.
Currently only branches and tags of the form `twisted-*` can use the `release` environment.
Only jobs from `.github/workflows/test.yaml` that are executed in the `release` environment can release to PyPI.
```

File: .github/workflows/test.yaml
Comment (adiroiban) [inline .github/workflows/test.yaml]:
```suggestion
  # Used exclusively for PyPi release publishing.
  # This is only executed on tags and only does the publishing.
  # The actual build is done by `release-build` job,
  # which is executed for each PR to help detect release
  # defects during the normal development cycle.
```

File: .github/workflows/test.yaml
Comment (adiroiban) [inline .github/workflows/test.yaml]:
```suggestion
      - release-build
```

File: .github/workflows/test.yaml
Comment (adiroiban) [inline .github/workflows/test.yaml]:
```suggestion
```

**Comment by adiroiban** [comment]
Thanks @graingert for the PR.

I will try to get this merge and then do a release using this code.

I have enabled auto-merge

I have removed the extra `release-publish-success-or-skipped:` job.

The thing is that the `all-succesful` job is designed to help with PRs checks.

The release job will alwasy be skipped for any PR.

I tried to add a comment/documentation for the `release-publish-success-or-skipped` job and I realised that we don't have a good use case of it.

**Comment by graingert** [comment]
Sounds good

**Comment by graingert** [comment]
What do the current PyPI trusted publishing settings look like?

**Comment by adiroiban** [comment]
For reference

The current PyPi configuration is

image

**Pull request merged**
